317 upload scores to query tag pw

philippmwirth opened the pull request:
# 317 uploade scores to query tag

Closes #317.

- exposed `labeled_set`, `unlabeled_set`, `query_set`, and `added_set` to user
- refactored the active learning agent
- upload active learning scores to query tag instead of preselected tag
- updated the docstrings and comments
- updated the documentation and tutorial
- updated the tests

The unmocked tests and the active learning tutorial run through on `api-dev` 
@ `commit 57bb1660402d54b953f05a3bf3562bf301ed63f6 `

IgorSusmelj commented on docs/source/getting_started/active_learning.rst:
```diff
@@ -100,11 +100,14 @@ Let's configure the sampling request and request an initial selection next:
    # we want an initial pool of 100 images
    config = SamplerConfig(n_samples=100, method=SamplingMethod.CORESET, name='initial-selection')
-   initial_selection = al_agent.query(config)
+   initial_selection, added_set = al_agent.query(config)
```
I don't like this example here. Would not use 100 for both. This just introduces confusion. Maybe start with 200 images and then add another 100. I also don't get the difference between initial selection and added set here

philippmwirth replied:
There is no difference in the initial step which is a little inelegant. Should we just go back to having a single return value? @MalteEbner thoughts?

IgorSusmelj replied:
Why should I get a tuple? That does not really make sense to me. I query --> want to get back what I specified in the query

MalteEbner replied:
I would go to have no return value at all, see my comments below.

IgorSusmelj reviewed (changes requested):
Tbh I'm a bit confused. I thought the idea is to just switch to where we upload the scores. This PR is changing quite a bit of the whole concept. We return a tuple? Why? We use sets instead of tag_ids? Why?
Could you please elaborate?

MalteEbner commented on lightly/active_learning/agents/agent.py:
```diff
@@ -34,78 +38,142 @@ class ActiveLearningAgent:
-    def __init__(self, api_workflow_client: ApiWorkflowClient, query_tag_name: str = None,
+    def __init__(self,
+                 api_workflow_client: ApiWorkflowClient,
+                 query_tag_name: str = None,
```
We could make it even easier and cleaner with `query_tag_name: str = "initial-tag"`. Then no checks are necessary if the `query_tag_name is None` and it is documented in code to which tag the query tag defaults.

IgorSusmelj replied:
Yes, could do that.

MalteEbner replied:
@philippmwirth What do you think about that?

philippmwirth commented:
> Tbh I'm a bit confused. I thought the idea is to just switch to where we upload the scores. This PR is changing quite a bit of the whole concept. We return a tuple? Why? We use sets instead of tag_ids? Why?
> Could you please elaborate?

The only conecptual change is that we upload the scores to the query tag.

- We always returned a tuple - the docs just didn't reflect this. See
https://github.com/lightly-ai/lightly/blob/7314e93077d11fc905c4a5152ad2ab4663122b22/lightly/active_learning/agents/agent.py#L155
- Where do we use sets instead of tag ids? We always had helpers to get the `labeled_set` and `unlabeled_set`, see 
https://github.com/lightly-ai/lightly/blob/7314e93077d11fc905c4a5152ad2ab4663122b22/lightly/active_learning/agents/agent.py#L69 I just refactored it and added the `query_set` so that users can easily upload predictions for the whole query set instead of the unlabeled set. The refactoring makes the code much easier to understand IMO.
- If you check out the changes in the tutorial or the unmocked test script, you can see that the only change from a user perspective is that the predictions should now be made on the `query_set` instead of the `unlabeled_set`.

Can you elaborate what your confusion is about?

MalteEbner commented:
> Tbh I'm a bit confused. I thought the idea is to just switch to where we upload the scores. This PR is changing quite a bit of the whole concept. We return a tuple? Why? We use sets instead of tag_ids? Why?
> Could you please elaborate?

We also switched that the scores are not computed for the `unlabeled_set` anymore, but for the whole `query_set`, thus the large number of changes in the active learning agent.

Furthermore, the functions in the active learning agent to compute the labeled_set, unlabeled_set, added_set and query_set out of the respective bitmasks became quite a lot and not that easily understandable, thus Philipp refactored that part to a much nicer structure.

I agree that we could rethink whether we really want to return a tuple. Now I would prefer to just return nothing, as the returned values are also easily accessible as attributes. It would make it unnecessary for the user to look up, what the returned values are.

MalteEbner commented on lightly/active_learning/agents/agent.py:
```diff
@@ -120,36 +188,40 @@ def query(self, sampler_config: SamplerConfig, al_scorer: Scorer = None) -> Tupl
+        self._preselected_tag_id = new_tag_data.id
+        self._preselected_tag_bitmask = self._get_preselected_tag_bitmask()
 
         return self.labeled_set, self.added_set
```
Perhaps we should just return nothing and let the user access the sets by him/herself. Then no documentation for the return values is need.
It is also a bit unlogical, that the query_set, which is needed for predicting the new scores, is not returned.

MalteEbner reviewed (approved):
I love your refactoring of the setting of the sets in the ActiveLearningAgent. I did not like the function `_set_labeled_and_unlabeled_set` I had implemented, but did not see an elegant way how to make it better. Now you found one which is great!
You also added further tests, which I always like :)

The comments I made are minor propositions, feel free to follow them or not.

IgorSusmelj commented:
> > Tbh I'm a bit confused. I thought the idea is to just switch to where we upload the scores. This PR is changing quite a bit of the whole concept. We return a tuple? Why? We use sets instead of tag_ids? Why?
> > Could you please elaborate?
> 
> The only conecptual change is that we upload the scores to the query tag.
> 
> * We always returned a tuple - the docs just didn't reflect this. See
>   https://github.com/lightly-ai/lightly/blob/7314e93077d11fc905c4a5152ad2ab4663122b22/lightly/active_learning/agents/agent.py#L155
> * Where do we use sets instead of tag ids? We always had helpers to get the `labeled_set` and `unlabeled_set`, see
>   https://github.com/lightly-ai/lightly/blob/7314e93077d11fc905c4a5152ad2ab4663122b22/lightly/active_learning/agents/agent.py#L69
>    I just refactored it and added the `query_set` so that users can easily upload predictions for the whole query set instead of the unlabeled set. The refactoring makes the code much easier to understand IMO.
> * If you check out the changes in the tutorial or the unmocked test script, you can see that the only change from a user perspective is that the predictions should now be made on the `query_set` instead of the `unlabeled_set`.
> 
> Can you elaborate what your confusion is about?

Let me test the PR first properly myself with the tutorial.

IgorSusmelj commented:
The other changes just confused me as I didn't expect any change besides where we store the scores.

Let's take the opportunity and streamline the workflow. There should be 1 way to do things. As Malte suggested, we could also drop the return value and streamline that the user just accesses the attribute of the al agent. I like that idea as it enforces consistency and simplifies the workflow.

IgorSusmelj commented:
The tutorials runs through also on my machine 👯

IgorSusmelj commented:
And thanks for all the clarifications

MalteEbner commented on lightly/active_learning/agents/agent.py:
```diff
@@ -14,14 +14,18 @@ class ActiveLearningAgent:
+            but not in the labeled set.
+        added_set:
+            Set of filenames corresponding to samples which were added to the 
+            labeled set in the last query.
```
Optional: "added to the labeled set and removed from the unlabeled set in the last query"

IgorSusmelj commented on docs/source/getting_started/active_learning.rst:
```diff
@@ -98,23 +98,26 @@ Let's configure the sampling request and request an initial selection next:
+   # we want an initial pool of 150 images
+   config = SamplerConfig(n_samples=150, method=SamplingMethod.CORESET, name='initial-selection')
+   al_agent.query(config)
+   initial_selection = al_agent.labeled_set
```
cold even drop this and directly to the print/ assert

IgorSusmelj reviewed (approved):
I like it a lot! I had some other ideas on how to improve but will summarize them in another issue :)
